### DISCUSSION TITLE
Report libraries that do and do not work!

### DISCUSSION THREAD

**Discussion by SpenceKonde** [SpenceKonde/DxCore#99] [Library discussion]
In the same spirit as the ongoing discussion on megaTinyCore... I know people must be using third party libraries with DxCore.

If people who are using this core could let me know which libraries they have used (include link to where that library comes from) or tried to use and whether they did or did not work, so I can gather all that together on a page with a big table of libraries, that'd be a big help, and we will all benefit from tables of known working libraries.

And also, a list of libraries known to not work (ideally, with a reason why they don't work so we can figure out what is needed to make them work and if it's worth the effort of doing.) would be of value too. 

It is highly likely that anything that works on megaTinyCore will work here. There were some rough patches w/regards to I2C in recent months (on both cores), but those are sorted out by now. As are some old SPI issues from the beginning of the year, and the cores are VERY similar. That list is at: https://github.com/SpenceKonde/megaTinyCore/blob/master/megaavr/extras/LibraryCompatibility.md

**Comment by FourDizzle** [comment]
I've been having trouble with getting the Ethernet lib to work with a w5500. Firstly, Udp.h needed the beginMulticast() virtual function added. But im still having issues. I think this might be a problem with the SPI lib but without and osciliscope or the any expertise with avr/arduino i can't speak intelligently. It might just be me.

**Reply by SpenceKonde** [reply]
Oh dear. Libraries for network connectivity support are also particularly important ones - particularly for the larger flash devices, like the Dx-series. 

My reflex and instinct is to hear that you had to touch udp.h and immediately wonder what else is wrong with the ArduinoAPI crap packaged with the core. 
Where did you pull the good definition of that from? Maybe we need to check for updates to those in general (of course, ArduinoAPI is not the same as "SpenceKondeAPI" which has been butchered to add stuff we needed and remove stuff that was bad and/or wrong from a design perspective. One can debate whether that was the right approach, so some finesse is needed here)

**Reply by FourDizzle** [reply]
I pulled the virtual function signature for beginMulticast from here https://github.com/arduino/ArduinoCore-avr/blob/master/cores/arduino/Udp.h. I was hoping an implementation existed somewhere, but at runtime my code didn't even get that far as it was unable to identify which model wiznet chip was connected. 

Either there is a more fundamental problem with the "SpenceKondeAPI"s SPI lib, or I'm fundamentally misunderstanding something basic (which is very possible). I'm still a bit new to embedded code and the avr architecture; could you point me to some working code that use's this core's SPI lib and possibly a resource that might shed some light on the scope of a "core" for the arduino ide? I haven't had enough exposure to the nitty gritty of the arduino ecosystem. I need to double check my thinking before I can meaningfully contribute to this codebase

**Reply by FourDizzle** [reply]
this is the branch im working in while I'm figuring this out. Not much different in it yet tho. https://github.com/FourDizzle/DxCore/tree/udp-multicast

**Comment by VernierCoven** [comment]
**DS1307RTC - works with small amendment**

This library almost works, but creates an instance which it calls RTC - this conflicts with the AVR's RTC. Renaming this instance within the library fixed it for me.

**Reply by SpenceKonde** [reply]
Where does that library come from?  Best would be github  link so I can sent him a PR that's just like 
```
#if (__AVR_ARCH__ > 100)
// these parts come with an on-chip RTC named RTC, so we need a different name
(code with the name modified to not conflict)
#else 
(current code)
#endif  
```

**Reply by VernierCoven** [reply]
It's [from here](https://github.com/PaulStoffregen/DS1307RTC).

There is a check in it to see if RTC is already defined, and if it is to undefine it. That was problematic for me, as I needed the AVR's RTC as well as the external RTC.

Cheers!

**Comment by SpenceKonde** [comment]
https://github.com/SpenceKonde/DS1307RTC
Does this work unmodified for you?

On Fri, Apr 30, 2021 at 3:22 PM VernierCoven ***@***.***>
wrote:

> It's from here <https://github.com/PaulStoffregen/DS1307RTC>.
>
> There is a check in it to see if RTC is already defined, and if it is to
> undefine it. That was problematic for me, as I needed the AVR's RTC as well
> as the external RTC.
>
> Cheers!
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/SpenceKonde/DxCore/discussions/99#discussioncomment-680893>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABTXEW6PWV2O6LK5IEF422TTLL7QRANCNFSM42P7IWGA>
> .
>


-- 
____________
Spence Konde
Azzy’S Electronics

New products! Check them out at tindie.com/stores/DrAzzy
GitHub: github.com/SpenceKonde
ATTinyCore <https://github.com/SpenceKonde/ATTinyCore>: Arduino support for
all pre-2016 tinyAVR with >2k flash!
megaTinyCore <https://github.com/SpenceKonde/megaTinyCore>: Arduino support
for all post-2016 tinyAVR parts!
DxCore <https://github.com/SpenceKonde/DxCore>: Arduino support for the AVR
Dx-series parts, the latest and greatest from Microchip!
Contact: ***@***.***

**Reply by VernierCoven** [reply]
Apologies - won't be able to test until next week. Will report back when I can.
Cheers.

**Comment by zmartins** [comment]
Hello, congratulations on your work.   
Is there any hope of finding compatible libraries for LiquidCrystal and  DRV8825 (StepperDriver) ?

**Reply by SpenceKonde** [reply]
Well, *I* need a version of LiquidCrystal that works for as project I'm doing (and I need to find or port one that supports not just i2c, but real 4-bit mode - I don't have the depth for the I2C backpack, and even if I did (I put a cutout for them in the first prototypes) the screen wouldn't be rigidly attached to the PCB (which is screwed down and hence is attached to the housing) and would need to be mounted somehow, and experience showed that hot-melt glue was not up to the task.  So it's much easier - not to mention cheaper -  for me to just use a part with more pins and talk to the part in 4 bit mode)  , so something needs to be done there... I've found one candidate library that some folks were talking up that was supposedly the one to use, though the architecture-independent code looked slow and shitty. Never actually wired it up to test

Can you link to the stepper driver in question? 
Have you tested it and found it to not work? I don't have any reports of someone having tried it?

**Reply by SpenceKonde** [reply]
If that's the StepperDriver in library manager, I would expect it to work as is, Have you tried it and found otherwise?

**Reply by zmartins** [reply]
I'm using the DRV8825 from https://github.com/laurb9/StepperDriver in a atmega2560 circuit that I would like to move to AVR128.   Not yet started, just trying to set up the project.

**Reply by SpenceKonde** [reply]
Yeah, that's the one. I would expect it to just work, 

> atmega2560 circuit that I would like to move to AVR128.

Good decision.

**Comment by zmartins** [comment]
About the LCD, I'm currently using https://github.com/laurb9/StepperDriver, the author says it is a derivate of the original Arduino LiquidCrystal Library but 5 times faster, and also works with 4, 8 wire parallel interface, I2C IO port expander (native I2C and bit bang) and Shift Regiter.  Maybe it fits your need.

**Comment by zmartins** [comment]
sorry, it is https://github.com/fmalpartida/New-LiquidCrystal

**Reply by SpenceKonde** [reply]
That library has a few general red flags - didn't get as far as lookinat whether rthere were specific issues that would impact you, but it looks nasty: 
1. It's OLD. How old? Old enough to predate the 1.6.x Arduino IDE versions (no library.properties). 
2. It does not appear to be actively maintained. 
3. There are three dozen files in the root folder - I can understand how someone might choose to not leap onto the new library format if it only had a few files, and everything was neaty organized and tidy.. That does not describe this library. They could have shoved all the mess into a src folder and kept everything clear, But they didn't. Because the maintainer if he still watches the repo, doesn't care. You always want to use software maintained by someone who cares, and who takes pride in their work.

The one I was looking at is this one: https://github.com/duinoWitchery/hd44780 There are some things that are less than impressive from a quick look - but it's got a decent reputation, and bperrybap is, as far as I am aware, of above average experience and ability within the Arduino community.

**Comment by ITDiscovery** [comment]
I believe PS2KeyAdvanced is working. I added the following to the PS2KeyAdvanced.h file:

```
// AVR128DB28
#if defined( DXCORE )
#define PS2_SUPPORTED           1
#define PS2_REQUIRES_PROGMEM    1
#define PS2_CLEAR_PENDING_IRQ   1
#endif
```

https://github.com/techpaul/PS2KeyAdvanced

I get similar output on an Arudino Pro Mini.
